File: examples/rails/app/graphql/mutations/update_user.rb
# frozen_string_literal: true

module Mutations
  class UpdateUser < BaseMutation
    argument :id, ID, required: true
    argument :email, String, required: false
    argument :name, String, required: false
    argument :password, String, required: false

    field :user, Types::UserType, null: false

    def resolve(id:, **)
      user = User.find_by(id:) || raise(GraphQL::ExecutionError, "User not found")
      user.update!(**)
      {user:}
    end
  end
end

File: examples/ts_apollo/src/resolvers/updateUser.ts
import { GraphQLError } from 'graphql';
import { MutationResolvers } from '../__generated__/graphql';
import _ from 'lodash';
import { generatePasswordDigest } from '../models/user.js';

const updateUser: MutationResolvers['updateUser'] = async (
  _parent,
  { input: { clientMutationId, id, email, name, password } },
  { authorize, db, serverSalt }
) => {
  await authorize();

  let user = await db.user.findUnique({ where: { id } });
  if (!user) throw new GraphQLError('User not found');

  const passwordDigest = password && generatePasswordDigest(serverSalt, password);
  const updatedAttributes = _.omitBy({ email, name, passwordDigest }, _.isUndefined);
  user = await db.user.update({ where: { id }, data: updatedAttributes });

  return { clientMutationId, user };
};

export default updateUser;
